src/comm.cc

```diff
@@ -1275,30 +1275,38 @@ commHandleWriteHelper(void * data)
     assert(clientInfo);
     assert(clientInfo->hasQueue());
     assert(clientInfo->hasQueue(queue));
-    assert(!clientInfo->selectWaiting);
     assert(clientInfo->eventWaiting);
     clientInfo->eventWaiting = false;
 
     do {
-        // check that the head descriptor is still relevant
-        const int head = clientInfo->quotaPeekFd();
-        Comm::IoCallback *ccb = COMMIO_FD_WRITECB(head);
+        if (clientInfo->writeOrQuotaDequeue())
+            return;
+    } while (clientInfo->hasQueue());
 
-        if (fd_table[head].clientInfo == clientInfo &&
-                clientInfo->quotaPeekReserv() == ccb->quotaQueueReserv &&
-                !fd_table[head].closing()) {
+    debugs(77,3, HERE << "emptied queue");
```

Comment:
I know you are not changing this line, but the diff implies otherwise, and it is likely that Amos will want you to change it because of that, so I would clean it up proactively:

```suggestion
    debugs(77,3, "emptied queue");
```

Your call though.

Reply:
Done.